### ISSUE TITLE
Loss function for all three scales

### ISSUE THREAD

**Issue by lukas-sl** [maiminh1996/YOLOv3-tensorflow#8]
I'm currently trying to implement YOLOv3 in TensorFlow, using the Estimator API. Your code and the graphics of the network structure have helped me a lot while doing this. However, I'm stuck at the loss function right now. I can't figure out, how to calculate the loss for all of the three scales. I've already looked at the paper and also tried to find the loss function in the darknet source code but can't figure it out.

Where did you find the equation/formula to use for this since anything I found is only for v1 or v2?

**Comment by SteveBetter** [comment]
```
        # K.binary_crossentropy is helpful to avoid exp overflow.
        xy_loss = object_mask * box_loss_scale * K.binary_crossentropy(raw_true_xy, raw_pred[..., 0:2],
                                                                       from_logits=True)
        wh_loss = object_mask * box_loss_scale * 0.5 * K.square(raw_true_wh - raw_pred[..., 2:4])
        confidence_loss = object_mask * K.binary_crossentropy(object_mask, raw_pred[..., 4:5], from_logits=True) + \
                          (1 - object_mask) * K.binary_crossentropy(object_mask, raw_pred[..., 4:5],
                                                                    from_logits=True) * ignore_mask
        class_loss = object_mask * K.binary_crossentropy(true_class_probs, raw_pred[..., 5:], from_logits=True)

        xy_loss = K.sum(xy_loss) / mf
        wh_loss = K.sum(wh_loss) / mf
        confidence_loss = K.sum(confidence_loss) / mf
        class_loss = K.sum(class_loss) / mf
        loss += xy_loss + wh_loss + confidence_loss + class_loss
```
I am looking for this too.
The  YOLOv3 paper says, [2.1 Bounding Box Prediction]"During training we use sum of squared error loss. If the ground truth for some coordinate prediction is t_stat_hat our gradient is the ground truth value (compared from the ground truth box) minus our prediction: t_star_hat - t_star. This ground truth value can be easily computed by inverting the equations above"
I also referred to the darknet source code. The x, y part seems to be sigmoid(tx_hat) - sigmoid(tx) and sigmoid(ty_hat) - sigmoid(ty). It's confusing.

Can anyone give the loss part a comment in detail? Thanks very much!

**Comment by SteveBetter** [comment]
I figured out how the darknet source code calculated the gradients.

sigmoid activation function -> cross_entropy loss:
L = - (y_hat * log(y) + (1 - y_hat) * log(1 - y))
dL / dy = (y - y_hat) / (y * (1 - y))
d(sigmoid(x)) / dx = y * (1 - y)
dL / dy * dy / dx = y - y_hat
The delta in darknet source code equals to -gradient = y_hat - y

linear activation - > sum of squared error loss:
L = 0.5(y_hat - y) ^2
dL / dy = - (y_hat - y)
dy / dx = 1
delta = y_hat - y
